main/src/ca/uwaterloo/flix/language/phase/sjvm/BytecodeCompiler.scala

```diff
@@ -17,34 +33,105 @@ object BytecodeCompiler {
+    case Expression.Unit(loc) =>
+      WithSource[R](loc) ~ pushUnit()
+
+    case Expression.Null(tpe, loc) =>
+      WithSource[R](loc) ~ pushNull()
+
+    case Expression.True(loc) =>
+      WithSource[R](loc) ~ pushBool(true)
+
+    case Expression.False(loc) =>
+      WithSource[R](loc) ~ pushBool(false)
+
+    case Expression.Char(lit, loc) =>
+      WithSource[R](loc) ~ pushChar(lit)
+
     case Expression.Float32(lit, loc) => ???
+
     case Expression.Float64(lit, loc) => ???
+
     case Expression.Int8(lit, loc) => ???
+
     case Expression.Int16(lit, loc) => ???
+
     case Expression.Int32(lit, loc) => ???
+
     case Expression.Int64(lit, loc) => ???
-    case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) => branch(compileExp(exp1), compileExp(exp2), compileExp(exp3))
+
+    case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+      branch(compileExp(exp1), compileExp(exp2), compileExp(exp3))
+
     case Expression.Ref(exp, tpe, loc) => ???
     //      NEW("class name") ~
     //      DUP ~
     //      compileExp(exp) ~
     //      INVOKESPECIAL("class name", "constructor signature")
+
     case Expression.Let(sym, exp1, exp2, tpe, loc) =>
-//      exp1.tpe match {
-//        case ErasedType.Int32() => WithSource[R](null) ~ compileExp(exp1) ~ popInt()
-//        case _ => ???
-//      }
-      compileExp(exp2)
+      // sym is unsafe. localvar stack + reg as types?
+      WithSource[R](loc) ~
+        compileExp(exp1) ~
+        XStore(sym, exp1.tpe) ~
+        compileExp(exp2)
+
+    case Expression.ArrayLoad(base, index, tpe, loc) =>
+      WithSource[R](loc) ~
+        compileExp(base) ~
+        compileExp(index) ~
+        ArrayInstructions.XALoad(tpe)
+
+    case Expression.ArrayStore(base, index, elm, tpe, loc) =>
+      WithSource[R](loc) ~
+        compileExp(base) ~
+        compileExp(index) ~
+        compileExp(elm) ~
+        ArrayInstructions.XAStore(elm.tpe) ~
+        pushUnit()
+
+    case Expression.ArrayLength(base, tpe, loc) =>
+      WithSource[R](loc) ~
+        compileExp(base) ~
+        ArrayInstructions.arrayLength()
+
+    case Expression.ArraySlice(base, beginIndex, endIndex, tpe, loc) =>
+      WithSource[R](loc) ~
+        compileExp(base) ~
+        compileExp(beginIndex) ~
+        compileExp(endIndex) ~
+        SWAP(beginIndex.tpe, endIndex.tpe) ~
+        DUP_X1(endIndex.tpe, beginIndex.tpe) ~
+        ISUB() ~
+        DUP()
+        // newArray(tpe) ~ not sure how to continue
+
+      ???
+
     case _ => ???
   }
 
   def WithSource[R <: Stack](loc: SourceLocation): F[R] => F[R] = ???
 
+  // TODO QUESTION: which functions are "unsafe" irt. types?
+
+  def ISWAP[R <: Stack](): F[R ** PrimInt32 ** PrimInt32] => F[R ** PrimInt32 ** PrimInt32] = ???
+
+  // TODO QUESTION: match on given types or copy per type?
+  def SWAP[R <: Stack, T <: JType, U <: JType](t: ErasedType[T], u: ErasedType[U]): F[R ** T ** U] => F[R ** U ** T] = (t, u) match {
```

Comment:
I would think this is better? Any reason to prefer the other one? 

We could also have both.